pkg/sql/sem/builtins/pgcrypto_builtins.go

```diff
@@ -116,3 +159,227 @@ var pgcryptoBuiltins = map[string]builtinDefinition{
+	}, nil
+}
+
+const (
+	minRoundsExtended = 1
+	maxRoundsExtended = 0xffffff
+)
+
+// generates a salt in the form _rrrrssss
+// "_" - xdes algorithm
+// "rrrr" - hashing cost (base64 encoded using itoa64) - number of rounds odd number between 1 inclusive and 16777215 inclusive
+// "ssss" - random salt (base64 encoded using itoa64)
+// see https://github.com/postgres/postgres/blob/586955dddecc95e0003262a3954ae83b68ce0372/contrib/pgcrypto/crypt-gensalt.c#L43
+func genSaltExtended(rounds int) ([]byte, error) {
+
+	if rounds == defaultRounds {
+		rounds = 29 * 25 // default
+	} else if rounds < minRoundsExtended || rounds > maxRoundsExtended || rounds%2 == 0 {
+		// Even iteration counts make it easier to detect weak DES keys from a look at the hash, so they should be avoided.
+		return nil, pgerror.Newf(
+			pgcode.InvalidParameterValue,
+			"Incorrect number of rounds %q. Must be %q or odd number between %q inclusive and %q inclusive.",
+			rounds, defaultRounds, minRoundsExtended, maxRoundsExtended)
+	}
+
+	output := make([]byte, 9)
+
+	output[0] = '_'
+	output[1] = itoa64[rounds&0x3f]
+	output[2] = itoa64[(rounds>>6)&0x3f]
+	output[3] = itoa64[(rounds>>12)&0x3f]
+	output[4] = itoa64[(rounds>>18)&0x3f]
+
+	input, err := getRandomBytes(3)
+	if err != nil {
+		return nil, err
+	}
+
+	value := uint32(input[0]) | uint32(input[1])<<8 | uint32(input[2])<<16
+	output[5] = itoa64[value&0x3f]
+	output[6] = itoa64[(value>>6)&0x3f]
+	output[7] = itoa64[(value>>12)&0x3f]
+	output[8] = itoa64[(value>>18)&0x3f]
+
+	return output, nil
+}
+
+// generates a salt in the form $1$ssssssss
+// "1" - md5 algorithm
+// "ssssssss" - random salt (base64 encoded using itoa64)
+// hashing cost is hard coded to 1000 rounds and does not appear in output
+// see https://github.com/postgres/postgres/blob/586955dddecc95e0003262a3954ae83b68ce0372/contrib/pgcrypto/crypt-gensalt.c#L79
+func genSaltMd5(rounds int) ([]byte, error) {
+
+	// can only be 1000
+	if err := checkFixedRounds(rounds, 1000); err != nil {
+		return nil, err
+	}
+
+	input, err := getRandomBytes(6)
+	if err != nil {
+		return nil, err
+	}
+
+	output := make([]byte, 11)
+
+	output[0] = '$'
+	output[1] = '1'
+	output[2] = '$'
+
+	value := uint32(input[0]) | uint32(input[1])<<8 | uint32(input[2])<<16
+	output[3] = itoa64[value&0x3f]
+	output[4] = itoa64[(value>>6)&0x3f]
+	output[5] = itoa64[(value>>12)&0x3f]
+	output[6] = itoa64[(value>>18)&0x3f]
+
+	value = uint32(input[3]) | uint32(input[4])<<8 | uint32(input[5])<<16
+	output[7] = itoa64[value&0x3f]
+	output[8] = itoa64[(value>>6)&0x3f]
+	output[9] = itoa64[(value>>12)&0x3f]
+	output[10] = itoa64[(value>>18)&0x3f]
+
+	return output, nil
+}
+
+const (
+	itoa64Blowfish    = "./ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+	minRoundsBlowfish = 4
+	maxRoundsBlowfish = 31
+)
+
+var encodingBlowfish = base64.NewEncoding(itoa64Blowfish).WithPadding(base64.NoPadding)
+
+// generates a salt in the form $2a$rr$ssssssssssssssssssssss
```

Comment:
nit: use `genSaltBlowfish generates ...`
same for all the above

Reply:
Done.